### ISSUE TITLE
Can ANTLR lexers/parsers be used for VS syntax highlighting and Error List?

### ISSUE THREAD

**Issue by stevecooperorg** [tunnelvisionlabs/antlr4cs#84]
Hi -- I'm using Antlr4cs and it's working beautifully; apologies if this is the wrong place to post!

I'm happily developing a language and I'd like to start integrating my language into an IDE. The natural place for me feels like integrating it into Visual Studio. It occurs to me that someone might have figured out a general way to plug an ANTLR lexer into the Visual Studio syntax highlighting system, or ANTLR parsers into an MSBuild task such that errors appear in the Visual Studio error list. 

Is there any kind of starting point, base package, plugin system, or similar which lets you integrate an ANTLR4CS language with Visual Studio's language support mechanisms?

Alternatively, has anyone had any success building, say, a Sublime Text build system / language def automatically from their grammar?

**Comment by timeraider4u** [comment]
Sorry for posting to an old thread, but I found this issue through Google search. The [language server protocol](http://langserver.org/) (or LSP for short) might be relevant to deal with these kind of questions. VS and other IDEs have already support for it. One drawback is that syntax highlighting is not supported yet by LSP (but it is in the backlog for the next version, see https://github.com/Microsoft/language-server-protocol/pull/124).

**Comment by timeraider4u** [comment]
Once the mentioned issue is resolved, I will try to implement a language server for the ANTLR4 grammar itself as proposed in https://github.com/antlr4ide/antlr4ide/issues/186

**Comment by SABIH-YUSSUF** [comment]
You can add syntax highlighting to a language using only a textMate file in VS Code. I am working on a language extension for a language for which there is an antlr2 parser, I don't know if there is a tool that allows converting the lexer rules in a  .g file to a textMate file. As to error lists, I think the best way to implement it is to write a language server that interacts with the parser

**Comment by timeraider4u** [comment]
Yes, you are right. As https://www.eclipse.org/community/eclipse_newsletter/2017/may/article3.php mentions, most LSP plugins for Eclipse also use TextMate for syntax highlighting. As Antlr can already generate code for syntax highlighting, I am not happy that syntax highlighting is not already supported through LSP because if it were, all dependencies could just be on the server side and the plug-ins for IDEs would get plain simple to implement. 

There currently is no converter for Antlr <-> TextMate grammars, as far as I know. There is a syntax highlighting grammar in TextMate for Antlr, but it seems to be an older version (Antlr2/Antlr3?): https://github.com/textmate/antlr.tmbundle

**Comment by timeraider4u** [comment]
Yes, you are right. As https://www.eclipse.org/community/eclipse_newsletter/2017/may/article3.php mentions, most LSP plugins for Eclipse also use TextMate for syntax highlighting. As Antlr can already generate code for syntax highlighting, I am not happy that syntax highlighting is not already supported through LSP because if it were, all dependencies could just be on the server side, the plug-ins for IDEs would get plain simple to implement and implementing syntax highlighting would not be much work.

There currently is no converter for Antlr <-> TextMate grammars, as far as I know. There is a syntax highlighting grammar in TextMate for Antlr, but it seems to be an older version (Antlr2/Antlr3?): https://github.com/textmate/antlr.tmbundle But even if it were Antlr version 4, it would require writing a converter anyway (and this is not something I would rather avoid if possible).
